py/dml/structure.py

```diff
@@ -2394,6 +2437,173 @@ def mkexpr(self, indices):
         return mkList(self.site,
                       [param.mkexpr(indices) for param in self.params])
 
+def port_class_ident(port):
+    '''The C identifier used for a port class within the device class
+    initialization function'''
+    return ('class' if port.objtype == 'device'
+            else '_port_class_' + port.attrname())
+
+def get_attr_name(prefix, node):
+    if node.objtype == 'register' and node.is_confidential():
+        return ctree.get_anonymized_name(node)
+    else:
+        return prefix + node.name
+
+
+class ConfAttrNameParamExpr(objects.ParamExpr):
+    '''The _attr_name parameter of an _conf_attribute object'''
+    __slots__ = ('site', 'node', 'cached',)
+
+    def __init__(self, site, node):
+        self.site = site
+        self.node = node
+        self.cached = None
+
+    def mkexpr(self, indices):
+        if self.cached is not None:
```

Comment:
Why not eager calculation? caching is a great root of evil, avoid it if you can.

Reply:
Note that the caching in ASTParamExpr is different: it saves a lot of time, and conditions for caching are complex, and they contain user-defined code, and the same param may give an error if evaluated too early, but be OK later on.

Reply:
I assume you're suggesting the eager calculation is to be done either in `make_autoparams` or in the `__init__`s of relevant ParamExprs?

I'm not entirely convinced regarding eager evaluation. Consider that my profiling does show that a few seconds are spent these on calculating these auto-optimized cached params. Though that number may be inflated by the profiling, I still wonder if we should risk spending unnecessary work if those params are left unused.

Besides, the complexity won't actually be reduced. All eager evaluation would do is to move all logic from `mkexpr` to `__init__`. Hell, you won't even eliminate the `cached` member as you need somewhere to store the result for later reference in `mkexpr` -- most you can do is rename the member and eliminate the `is not None` check.

Reply:
I agree that the damage is limited (the main cost is readability, the reader of those `ParamExpr`s doesn't understand why caching is needed). But the wins are also limited if not nonexistent, most of these ParamExprs exist because we know that they (almost) always are read at least once by dml-builtins, right?

So I cite Occam, please remove caching because why not remove it.

Reply:
We've debated back and forth about this off-github. We've decided on keeping the caching, but adding a blurb to explain caching should be used with care.